pkg/virt-api/webhooks/s390x.go

```diff
@@ -0,0 +1,62 @@
+/* Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright the KubeVirt Authors.
+ *
+ */
+
+/*
+ * s390x utilities are in the webhooks package because they are used both
+ * by validation and mutation webhooks.
+ */
+package webhooks
+
+import (
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/pointer"
+)
+
+func setDefaultS390xDisksBus(spec *v1.VirtualMachineInstanceSpec) {
+	bus := v1.DiskBusVirtio
+
+	for i := range spec.Domain.Devices.Disks {
+		disk := &spec.Domain.Devices.Disks[i].DiskDevice
+
+		if disk.Disk != nil && disk.Disk.Bus == "" {
+			disk.Disk.Bus = bus
+		}
+		if disk.CDRom != nil && disk.CDRom.Bus == "" {
+			disk.CDRom.Bus = bus
+		}
+		if disk.LUN != nil && disk.LUN.Bus == "" {
+			disk.LUN.Bus = bus
+		}
+	}
+}
+
+// Disable ACPI Feature by default on s390x, since it is not supported
+func setS390xDefaultFeatures(spec *v1.VirtualMachineInstanceSpec) {
+	featureStateDisabled := v1.FeatureState{Enabled: pointer.P[bool](false)}
+	if spec.Domain.Features == nil {
+		spec.Domain.Features = &v1.Features{
+			ACPI: featureStateDisabled,
+		}
+	} else if spec.Domain.Features.ACPI.Enabled == nil {
+		spec.Domain.Features.ACPI.Enabled = pointer.P[bool](false)
```

Comment:
not sure here if we should return an error instead of silently disable it. Does the webhook reject this option for s390x?

Reply:
This only sets it to disabled on s390x by default. You can still explicitly enable it.
Though it does not work, so your VM will then fail with an error in the logs about it.
Not sure what the design here is, if the api should verify that the resulting livbirt xml is sane for the arch or leaving that to libvirt.

Reply:
Ok, make sense. In general the XML is verified by libvirt, however, the webhook should reject the wrong option set on the VM

Reply:
I had the same question while looking at this code, and I'm not sure the above really answers it.

Why is it necessary to explicitly disable the feature? Is it not enough to leave its status unspecified and rely on libvirt doing the right thing?

Reply:
I think it makes sense for the api to reject options that can't be translated to libvirt XML, but the rest of the verification should be done by libvirt, as it can also be node specific, so i don't know if the API could even verify it properly, as it can't know the specific node the VM will be scheduled on. In this specific case it is dependent on the architecture and libvirt version (as older versions ignore ACPI if it is enabled but does not work). Other cases could be a feature dependent on a specific CPU generation.